linera-views/README.md

```diff
@@ -11,6 +11,134 @@ atomically persisted in storage).
+also need to accomodate other features:
+* Administrative data (like journal) need to store some data starting from the `base_key`.
+* Some of the values need to be split into several smaller values.
+
+The rule for constructing are the following:
+* For the construction of `struct` objects of associated base key `bk` we do the
+following: If the corresponding type has entries `entry_0`, ..., `entry_k` then the
+base key of the object associated to the `k`-th entry is `[bk * * * *]` with `[* * * *]`
+the serialization of the number `k` considered as a `u32`.
+* For each view type, we have some associated data. Some are the hash, others the counts
+(e.g. for `LogView` and `QueueView`) and some the index. The corresponding enum is
+named `KeyTag` for each of those containers. For the index, the values are serialized.
+* For a given base key `base_key` the keys of the form `[base_key tag]` with tag in
+`0..MIN_VIEW_TAG` are reserved for administrative data.
+
+These construction are done in order to satisfay the following property: The set of keys
+of a view is prefix-free.
+
+
+## Chosen implementation
+
+# Implementation of the `keyValueStoreClient`.
+
+The design of the key value store client is done in the following way:
+* We can put a `(key, value)` and delete a `key`.
+* We can retrieve the list of keys having a specific prefix. Similarly, we can
+retrieve the list of `(key, value)` whose key have a specific prefix.
+* Also, the list of the keys are provided in the lexicographic order.
+
+# The journaling.
+
+We want to make sure that all operations can be done on the database in a sure
+way. That is the operation have to be processed completely. The idea is to
+store the journal in the `[base_key 0 * * * *]` keys with `[* * * *]` the
+serialization of an `u32` entry. Once written, the journal is processed
+until all entries have been treated.
+Those keys are not overlaping with the other keys since their value is of the
+form `[base_key tag *]` with `tag >= MIN_VIEW_TAG` and currently we have
+`MIN_VIEW_TAG = 1`.
+
+
+## The `KeyValueStoreClientBigValue` type.
+
+Each key value store client has a maximal size of its values (named `MAX_VALUE_SIZE`
+in the code). If we want to have unlimited size, then we can build a
+key-value store client with unlimited size.
+
+The built key-value store client has `MAX_VALUE_SIZE` set to its maximum possible
+value. This forces us to split the value into several smaller values. One key
+such example is `DynamoDb` which has a limit of 400kB on its value size.
+
+Design is the following:
+* For every `key` we build several corresponding keys `[key * * * *]` that contain
+each a piece of the full data set.
+The `[* * * *]` is the serialization of the index which is `u32` so that
+the order of the serialization is the same as the order of the `u32` entries.
+* For the first key, that is `[key 0 0 0 0]` the information is `[* * * * value_0]`
```

Comment:
It's confusing to use terms like "data set", "information" and "value" interchangeably. Maybe we should establish a terminology instead to distinguish between the large, logical, high-level value, and the small, low-level, stored values, before explaining how one the former gets split up into the latter.

Reply:
Ok, the "information" and "data set" have been removed.
Regarding the large, logical, high-value, I am not completely sure what you mean and maybe we do not need such terminology.

Reply:
I mean the large value, that we can't store.
As opposed to the small values that we do store and that represent the large value's chunks.